File: apps/server/src/routes/transaction.ts
import { Router, Request, Response, NextFunction } from "express";
import { deviceAuthMiddleware } from "../middleware/deviceAuth";
import { validate } from "../middleware/validate";
import { transferSchema } from "../validators/transaction";
import * as transactionService from "../services/transactionService";
import prisma from "../config/prisma";

const router = Router();

const asyncHandler =
  (fn: (req: Request, res: Response, next: NextFunction) => Promise<void>) =>
  (req: Request, res: Response, next: NextFunction) =>
    fn(req, res, next).catch(next);

router.use(deviceAuthMiddleware);

router.post(
  "/transfer",
  validate(transferSchema),
  asyncHandler(async (req: Request, res: Response) => {
    const tx = await transactionService.transfer(req.body, req.device!.deviceId);
    res.status(201).json(tx);
  })
);

router.get("/", asyncHandler(async (req: Request, res: Response) => {
  const walletId = req.query.walletId as string;
  if (!walletId) {
    res.status(400).json({ error: "walletId query parameter is required" });
    return;
  }

  // 权限校验：验证设备是否关联该钱包
  const subscription = await prisma.walletSubscription.findFirst({
    where: {
      wallet_id: walletId,
      device_id: req.device!.deviceId,
    },
  });
  if (!subscription) {
    // 兜底：钱包存在即允许（刚创建还没添加网络账户）
    const wallet = await prisma.wallet.findUnique({ where: { id: walletId } });
    if (!wallet) {
      res.status(403).json({ error: "You do not have permission to view this wallet's transactions" });
      return;
    }
  }

  const page = parseInt((req.query.page as string) || "1", 10);
  const limit = Math.min(parseInt((req.query.limit as string) || "20", 10), 100);

  // Parse filter params
  const type = req.query.type as string | undefined;
  const timeRange = req.query.timeRange as string | undefined;
  const search = req.query.search as string | undefined;
  const tokenSymbol = req.query.tokenSymbol as string | undefined;

  const result = await transactionService.getTransactions({
    walletId,
    page,
    limit,
    type: (type === "send" || type === "receive") ? type : undefined,
    timeRange: (["today", "7d", "30d", "90d"].includes(timeRange || "")) ? timeRange as any : undefined,
    search: search || undefined,
    tokenSymbol: tokenSymbol || undefined,
  });
  res.json(result);
}));

/**
 * GET /transactions/check-address?address=xxx
 * 查询地址是否在系统中（contacts 在客户端，不再查询）
 */
router.get("/check-address", asyncHandler(async (req: Request, res: Response) => {
  const address = (req.query.address as string || "").trim();
  if (!address) {
    res.status(400).json({ error: "address 参数不能为空" });
    return;
  }

  // 查系统内：wallets_addresses 表（所有链地址）
  const walletAddress = await prisma.walletAddress.findFirst({ where: { address } });
  const inSystem = !!walletAddress;

  // contacts 在客户端，inContacts 始终返回 false
  res.json({ inSystem, inContacts: false });
}));

router.get("/:id", asyncHandler(async (req: Request, res: Response) => {
  const txId = req.params.id as string;
  const tx = await transactionService.getTransactionDetail(txId);
  res.json(tx);
}));

export default router;

File: apps/server/src/routes/transaction.rs
//! 交易路由 — /api/v1/transactions
//! 迁移自 IMWallet routes/transaction.ts (4 个接口)

use crate::errors::AppError;
use crate::middleware::{AppState, DevicePayload};
use crate::models::Transaction;
use crate::services::transaction_service;
use axum::{
    extract::{Path, Query, State},
    routing::{get, post},
    Extension, Json, Router,
};
use serde::{Deserialize, Serialize};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/transactions/transfer", post(transfer))
        .route("/transactions", get(get_transactions))
        .route("/transactions/check-address", get(check_address))
        .route("/transactions/{id}", get(get_transaction))
}

/// POST /transactions/transfer — 执行转账
async fn transfer(
    State(state): State<AppState>,
    Extension(device): Extension<DevicePayload>,
    Json(body): Json<transaction_service::TransferInput>,
) -> Result<
    (
        axum::http::StatusCode,
        Json<transaction_service::TransferResult>,
    ),
    AppError,
> {
    let result = transaction_service::execute_transfer(
        state.db.clone(),
        body,
        &device.device_id,
        &device.platform,
        &state.config,
    )
    .await?;
    Ok((axum::http::StatusCode::CREATED, Json(result)))
}

#[derive(Debug, Deserialize)]
pub struct TransactionQuery {
    pub wallet_id: String,
    pub page: Option<u64>,
    pub limit: Option<u64>,
    #[serde(default)]
    pub token_symbol: Option<String>,
}

#[derive(Debug, Serialize)]
struct TransactionsResponse {
    transactions: Vec<TransactionItem>,
    total: u64,
}

/// 交易列表返回 DTO — status 写死 CONFIRMED（私有链即时确认，无状态流转）
#[derive(Debug, Serialize)]
struct TransactionItem {
    id: String,
    from_address: String,
    to_address: String,
    token_symbol: String,
    amount: rust_decimal::Decimal,
    fee: rust_decimal::Decimal,
    status: String,
    memo: String,
    platform: String,
    created_at: Option<rbdc::DateTime>,
}

impl From<Transaction> for TransactionItem {
    fn from(t: Transaction) -> Self {
        Self {
            id: t.id,
            from_address: t.from_address,
            to_address: t.to_address,
            token_symbol: t.token_symbol,
            amount: t.amount,
            fee: t.fee,
            status: "CONFIRMED".to_string(),
            memo: t.memo,
            platform: t.platform,
            created_at: t.created_at,
        }
    }
}

async fn get_transactions(
    State(state): State<AppState>,
    Query(query): Query<TransactionQuery>,
) -> Result<Json<TransactionsResponse>, AppError> {
    let page = query.page.unwrap_or(1);
    let limit = query.limit.unwrap_or(20);
    let (txns, total) = transaction_service::get_transactions(
        state.db.clone(),
        &query.wallet_id,
        query.token_symbol.as_deref(),
        page,
        limit,
    )
    .await?;

    Ok(Json(TransactionsResponse {
        transactions: txns.into_iter().map(TransactionItem::from).collect(),
        total,
    }))
}

#[derive(Debug, Serialize)]
struct CheckAddressResponse {
    in_system: bool,
}

/// GET /transactions/check-address?address=xxx — 检查地址是否在系统内
#[derive(Debug, Deserialize)]
pub struct CheckAddressQuery {
    pub address: String,
}

async fn check_address(
    State(state): State<AppState>,
    Query(query): Query<CheckAddressQuery>,
) -> Result<Json<CheckAddressResponse>, AppError> {
    let in_system = transaction_service::check_address(state.db.clone(), &query.address).await?;
    Ok(Json(CheckAddressResponse { in_system }))
}

/// GET /transactions/:id — 获取交易详情
async fn get_transaction(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<TransactionItem>, AppError> {
    let txn = transaction_service::get_transaction(state.db.clone(), &id)
        .await?
        .ok_or_else(|| AppError::NotFound("交易不存在".into()))?;
    Ok(Json(TransactionItem::from(txn)))
}
